charmhelpers/contrib/network/files/check_sriov_numvfs.py

```diff
@@ -0,0 +1,128 @@
+SRIOV_NUMVFS_TEMPLATE = "/sys/class/net/{0}/device/sriov_numvfs"
+SRIOV_TOTALVFS_TEMPLATE = "/sys/class/net/{0}/device/sriov_totalvfs"
+
+
+def get_interface_setting(file):
+    """Return the value content of a setting file as int"""
+    with open(file) as f:
+        value = f.read()
+    return int(value)
+
+
+def check_interface_numvfs(iface, numvfs):
+    """Check SR-IOV numvfs config"""
+    sriov_numvfs_path = SRIOV_NUMVFS_TEMPLATE.format(iface)
+    sriov_totalvfs_path = SRIOV_TOTALVFS_TEMPLATE.format(iface)
+    msg = []
+
+    # Ignore non-existing interfaces
+    if not os.path.exists(DEVICE_TEMPLATE.format(iface)):
+        return []
+
+    # Ensure that SR-IOV/VT-d and IOMMU are enabled
+    if (not os.path.exists(sriov_totalvfs_path) or
+            not os.path.exists(sriov_numvfs_path)):
+        return ["{}: VFs are disabled or not-available".format(iface)]
+
+    # Verify number of virtual functions
+    sriov_numvfs = get_interface_setting(sriov_numvfs_path)
+    sriov_totalvfs = get_interface_setting(sriov_totalvfs_path)
+    if numvfs != sriov_numvfs:
+        msg.append(
+            "{}: Number of VFs on interface ({}) does not match expected ({})"
+            .format(iface, sriov_numvfs, numvfs)
+        )
+    if numvfs > sriov_totalvfs:
+        msg.append(
+            "{}: Maximum number of VFs available on interface ({}) is lower than the expected ({})"
+            .format(iface, sriov_totalvfs, numvfs)
+        )
+    return msg
+
+
+def parse_sriov_numvfs(device_numvfs):
+    """Parse parameters and check format"""
+    msg = "Parameter format must be '<interface>:<numvfs>', e.g. ens3f0:32, given: {}".format(device_numvfs)
+    assert device_numvfs != '', msg
+    parts = device_numvfs.split(':')
+    assert len(parts) == 2, msg
+    assert len(parts[0]) > 0, msg
+    assert int(parts[1]) > 0, msg
+    iface = str(device_numvfs.split(':')[0])
+    numvfs = int(device_numvfs.split(':')[1])
+    return (iface, numvfs)
+
+
+def parse_args():
+    """Parse command-line options."""
+    parser = argparse.ArgumentParser(description="Check SR-IOV number of virtual functions configuration")
+    parser.add_argument("sriov_numvfs", nargs='+', help="format: <interface>:<numvfs>")
+    args = parser.parse_args()
+    return args
+
+
+def main():
+    """Parse args and check the sriov numvfs."""
+    args = parse_args()
+    error_msg = []
+
+    try:
+        for device_numvfs in args.sriov_numvfs:
+            iface, numvfs = parse_sriov_numvfs(device_numvfs)
+            error_msg += check_interface_numvfs(iface, numvfs)
+        if error_msg:
+            print("CRITICAL: {} problems detected\n".format(len(error_msg)) + "\n".join(error_msg))
+            sys.exit(2)
+    except:  # noqa: E722
+        print("{} raised unknown exception '{}'".format(__file__, sys.exc_info()[0]))
+        traceback.print_exc(file=sys.stdout)
+        sys.exit(3)
+
+    print("OK: sriov_numvfs set to " + ", ".join(args.sriov_numvfs))
```

Comment:
nitpick: Use `"... {}".format(...)` as in the print within the except scope (and other places) for readability.

Reply:
done